File: frontend/src/types/portfolio.ts
export interface Project {
    id: number;
    github_id: string;
    name: string;
    display_name: string;
    description: string | null;
    github_url: string;
    live_url: string | null;
    primary_language: string | null;
    languages: Record<string, number>;
    topics: string[];
    stars_count: number;
    forks_count: number;
    watchers_count: number;
    is_forked: boolean;
    is_featured: boolean;
    category: string | null;
    github_created_at: string | null;
    github_updated_at: string | null;
}

export interface ProjectListResponse {
    projects: Project[];
    total_count: number;
    featured_count: number;
}

export interface Skill {
    id: number;
    name: string;
    category: string;
    proficiency: number;
    icon_name: string | null;
    color: string | null;
    years_experience: number | null;
    is_highlighted: boolean;
}

export interface SkillsGrouped {
    languages: Skill[];
    frameworks: Skill[];
    databases: Skill[];
    tools: Skill[];
    cloud: Skill[];
    other: Skill[];
}

export interface Experience {
    id: number;
    company_name: string;
    role_title: string;
    description: string | null;
    highlights: string[];
    start_date: string;
    end_date: string | null;
    is_current: boolean;
    location: string | null;
    company_logo_url: string | null;
}

export interface Certificate {
    id: number;
    title: string;
    issuer: string;
    issue_date: string;
    credential_url: string | null;
    credential_id: string | null;
}

export interface ProfileInfo {
    full_name: string;
    title: string;
    tagline: string;
    location: string;
    email: string;
    phone: string;
    github_url: string;
    linkedin_url: string;
    avatar_url: string | null;
    resume_url: string | null;
    about_short: string;
    about_long: string | null;
    is_available_for_hire: boolean;
}

export interface PortfolioStats {
    total_projects: number;
    total_commits: number;
    primary_languages: Array<{ name: string; count: number }>;
    language_distribution: Record<string, number>;
    total_stars: number;
    total_forks: number;
    most_active_project: string | null;
    projects_by_category: Record<string, number>;
    skills_count: number;
    experience_years: number;
    certifications_count: number;
}

export interface ContactFormData {
    sender_name: string;
    sender_email: string;
    subject: string;
    message_body: string;
    company_name?: string;
}

export interface ContactSubmitResponse {
    success: boolean;
    message: string;
    reference_id: number;
}

File: backend/app/schemas/portfolio.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field, EmailStr, field_validator
from enum import Enum


class ProjectCategory(str, Enum):
    FULL_STACK = "full_stack"
    BACKEND = "backend"
    FRONTEND = "frontend"
    ML_AI = "ml_ai"
    IOT = "iot"
    DSA = "dsa"
    PERSONAL = "personal"


class ProjectResponse(BaseModel):
    id: Optional[int] = None
    github_id: str
    name: str
    display_name: str
    description: Optional[str] = None
    github_url: str
    live_url: Optional[str] = None
    primary_language: Optional[str] = None
    languages: dict[str, float] = Field(default_factory=dict)
    topics: list[str] = Field(default_factory=list)
    stars_count: int = 0
    forks_count: int = 0
    is_forked: bool = False
    is_featured: bool = False
    category: Optional[str] = None
    
    class Config:
        from_attributes = True


class ProjectListResponse(BaseModel):
    projects: list[ProjectResponse]
    total_count: int


class SkillResponse(BaseModel):
    name: str
    category: str
    proficiency: int
    color: Optional[str] = None


class SkillsGroupedResponse(BaseModel):
    languages: list[SkillResponse]
    frameworks: list[SkillResponse]
    databases: list[SkillResponse]
    tools: list[SkillResponse]
    cloud: list[SkillResponse]


class ExperienceResponse(BaseModel):
    company_name: str
    role_title: str
    description: Optional[str] = None
    highlights: list[str] = Field(default_factory=list)
    start_date: str
    end_date: Optional[str] = None
    is_current: bool = False
    location: Optional[str] = None


class CertificateResponse(BaseModel):
    title: str
    issuer: str
    issue_date: str
    credential_url: Optional[str] = None


class PortfolioStatsResponse(BaseModel):
    total_projects: int
    total_stars: int
    total_forks: int
    primary_languages: list[dict]
    visitors_count: int
    messages_count: int


class ProfileInfoResponse(BaseModel):
    full_name: str = "Ashish Prasad Gupta"
    title: str = "Full Stack Developer & CS Student"
    tagline: str = "Building the future, one commit at a time"
    location: str = "New Delhi, India"
    email: str = "ashishguptaop195@gmail.com"
    phone: str = "+91 9667233973"
    github_url: str = "https://github.com/Aashish-Op"
    linkedin_url: str = "https://linkedin.com/in/ashish-gupta"
    about_short: str = "Passionate Computer Science student at LPU with expertise in full-stack development, IoT, and machine learning."
    is_available_for_hire: bool = True


class ContactMessageCreate(BaseModel):
    sender_name: str = Field(..., min_length=2, max_length=255)
    sender_email: EmailStr
    subject: str = Field(..., min_length=5, max_length=500)
    message_body: str = Field(..., min_length=20, max_length=5000)
    company_name: Optional[str] = Field(None, max_length=255)
    
    @field_validator("sender_name")
    @classmethod
    def validate_name(cls, v: str) -> str:
        return v.strip()
    
    @field_validator("subject", "message_body")
    @classmethod
    def strip_whitespace(cls, v: str) -> str:
        return v.strip()


class ContactSubmitResponse(BaseModel):
    success: bool
    message: str
    reference_id: int


class UserResponse(BaseModel):
    id: int
    clerk_id: str
    email: str
    name: str
    avatar_url: Optional[str] = None
    created_at: datetime
    
    class Config:
        from_attributes = True


class VisitorStatsResponse(BaseModel):
    total_visitors: int
    unique_visitors: int
    authenticated_visitors: int
    recent_visitors: list[dict]


class HealthCheckResponse(BaseModel):
    status: str
    version: str
    database: str
    timestamp: datetime
